src/utils/logging.jl

```diff
@@ -0,0 +1,188 @@
+
+import Logging
+
+export open_file_logger
+export MultiLogger
+export LogEvent
+export LogEventTracker
+export report_log_summary
+export get_log_events
+
+
+"""
+    open_file_logger(func, filename[, level, mode])
+
+Opens a file logger using Logging.SimpleLogger.
+
+# Example
+```julia
+open_file_logger("log.txt", Logging.Info) do logger
+    global_logger(logger)
+    @info "hello world"
+end
+```
+"""
+function open_file_logger(func::Function, filename::String, level=Logging.Info, mode="w+")
+    stream = open(filename, mode)
+    try
+        logger = Logging.SimpleLogger(stream, level)
+        func(logger)
+    finally
+        close(stream) 
+    end
+end
+
+"""Contains information describing a log event."""
+mutable struct LogEvent
+    file::String
+    line::Int
+    id::Symbol
+    message::String
+    level::Logging.LogLevel
+    count::Int
+    suppressed::Int
+end
+
+LogEvent(file, line, id, message, level) = LogEvent(file, line, id, message, level, 1, 0)
+
+struct LogEventTracker
+    events::Dict{Logging.LogLevel, Dict{Symbol, LogEvent}}
+
+    LogEventTracker(events::Dict{Logging.LogLevel, Dict{Symbol, LogEvent}}) = new(events)
```

Comment:
Is this inner constructor necessary?

Reply:
Without it I get a method overwrite error for the default constructor that takes Any.  I'm choosing to leave it for now.  I'm open to discussion, though.

Reply:
In this sort of a situation, I would add a comment for why a inner constructor is necessary.

Reply:
Good point, added.